TST: Fix bugs in various tests found via linter

nickodell opened the pull request:
<!-- 
Thanks for contributing a pull request! Please ensure that
your PR satisfies the checklist before submitting:
http://scipy.github.io/devdocs/dev/contributor/development_workflow.html#checklist-before-submitting-a-pr

Also, please name and describe your PR as you would write a
commit message:
http://scipy.github.io/devdocs/dev/contributor/development_workflow.html#writing-the-commit-message.
However, please only include an issue number in the description, not the title,
and please ensure that any code names containing underscores are enclosed in backticks.

Depending on your changes, you can skip CI operations and save time and energy: 
http://scipy.github.io/devdocs/dev/contributor/continuous_integration.html#skipping

Note that we are a team of volunteers; we appreciate your
patience during the review process.

Again, thanks for contributing!
-->

#### Reference issue
<!--Example: Closes gh-WXYZ.-->
No issue was filed.

#### What does this implement/fix?
I was looking through the flake bugbear tests to see if any of them have value to the SciPy project. I found some warnings in the tests which look like bugs to me.

Specifically, I was looking at:

* B007: Loop control variable {name} not used within loop body
* B015: Pointless comparison. Did you mean to assign a value? Otherwise, prepend assert or remove it.
* B017: pytest.raises(Exception) should be considered evil

Many of the B007 warnings look like false positives, but some of them look like clear bugs to me. For example, in the test for kendalltau, a comment says that both variants of the kendalltau test should give the same result, and there is a loop over variants, but the variant is not passed to the function, so it just runs the same test twice.

For B015, there are tests which execute a statement like `a == b`, but do nothing with the result of that comparison. This looks like a bug to me, unless the intent was just to check that the equality check runs without error, and the result of the equality check is unimportant.

For B017, I found one place in the tests where a more specific exception can be expected, which makes the test more specific.

#### Additional information

I found one other thing that looks suspicious.

This test loops over two values of y: 0, and a random value. However, the variable y is never used. This looks like a bug to me. Either this shouldn't be a loop, or y should be involved in the check somehow. However, I'm unsure what the fix is.

```
    def test_roots_croots(self):
        # Test the complex root finding algorithm
        np.random.seed(1234)

        for k in range(1, 15):
            c = np.random.rand(k, 1, 130)

            if k == 3:
                # add a case with zero discriminant
                c[:,0,0] = 1, 2, 1

            for y in [0, np.random.random()]:
                w = np.empty(c.shape, dtype=complex)
                _ppoly._croots_poly1(c, w)

                if k == 1:
                    assert_(np.isnan(w).all())
                    continue

                res = 0
                cres = 0
                for i in range(k):
                    res += c[i,None] * w**(k-1-i)
                    cres += abs(c[i,None] * w**(k-1-i))
                with np.errstate(invalid='ignore'):
                    res /= cres
                res = res.ravel()
                res = res[~np.isnan(res)]
                assert_allclose(res, 0, atol=1e-10)
```

https://github.com/scipy/scipy/blob/main/scipy/interpolate/tests/test_interpolate.py#L1623

nickodell committed:
TST: Check with mmap both enabled and disabled

This test appears to have been intended to run twice: with and without
mmap.

nickodell committed:
TST: stats: check all variants of kendalltau

nickodell committed:
TST: linalg: check that all lapack drivers raise error

nickodell committed:
TST: linalg: Remove loop over A because it has no influence on test

nickodell committed:
TST: linalg: require specific exception type

nickodell committed:
TST: differentiate: assert that values are equal

nickodell committed:
TST: sparse: check that dunder methods in reverse are equal

nickodell committed:
TST: differentiate: check feval matches for numpy

mdhaber commented:
re: the other suspicious one, it's possible it might have started with a plan to treat it like `test_roots_random`, but that was abandoned? @ev-br I see this was a long time ago, but does this need to use `y`?
https://github.com/scipy/scipy/commit/19b3f3ed3085750812d235f56032b3c4be9de83c#diff-3dc1d145e06c5b57ea93006ccb5028bd6d8bb51d2738c575583858868ab6bb4f

ev-br commented:
Indeed, good catch! The loop can remove indeed.

mdhaber commented:
@ev-br it looks like `y` was added as an argument of `_croots_poly1` in that PR, but I don't see another test for it in that file. Should the loop be removed, or can we add `y` into the test somehow?

ev-br commented:
`croots_poly1` is used by `_ppoly.real_roots`, which is in turn used by `PPoly.solve`. So in effect `croots_poly1` is tested in `test_roots_random`. 

Alternatively, we can keep the test with non-zero `y` and fix the test:

```diff
$ git diff
diff --git a/scipy/interpolate/tests/test_interpolate.py b/scipy/interpolate/tests/test_interpolate.py
index 9cf34c8b90..0ee5f45be9 100644
--- a/scipy/interpolate/tests/test_interpolate.py
+++ b/scipy/interpolate/tests/test_interpolate.py
@@ -1622,13 +1622,13 @@ class TestPPoly:
 
             for y in [0, np.random.random()]:
                 w = np.empty(c.shape, dtype=complex)
-                _ppoly._croots_poly1(c, w)
+                _ppoly._croots_poly1(c, w, y)
 
                 if k == 1:
                     assert_(np.isnan(w).all())
                     continue
 
-                res = 0
+                res = -y
                 cres = 0
                 for i in range(k):
                     res += c[i,None] * w**(k-1-i)
```

The test IIUC checks that the value of the polynomial is close to zero at what is returned as the root. For non-zero `y`, `croots_poly1` computes the solution to `f(x) = y` instead of `f(x) = 0`. Thus need to check that `f(x) - y` is close to zero.

mdhaber commented:
Thanks! I see. Your choice @nickodell but since the loop is already there, I think I'd roll with it.

nickodell committed:
TST: test solving roots with zero and random y

Due to a bug, this test ignores the y value in this loop, and
only checks that the root could be solved with y=0. Fix this.

Co-authored-by: Evgeni Burovski <evgeny.burovskiy@gmail.com>

nickodell commented:
>Alternatively, we can keep the test with non-zero `y` and fix the test:

Thanks. I've implemented this suggestion.

lucascolley commented:
CI failures unrelated, gh-20806 and gh-21311. thanks @nickodell !
